lib/streamlit/elements/multiselect.py

```diff
@@ -60,14 +67,14 @@ def multiselect(
         kwargs: Optional[WidgetKwargs] = None,
         *,  # keyword-only arguments:
         disabled: bool = False,
-    ) -> List[T]:
+    ) -> List[str]:
```

Comment:
@vdonato This is probably the biggest surprise I had while working on this PR. Kinda weird that enums are treated differently than any other type by multiselect. Not sure this is intentional.

Reply:
Argh -- while I wouldn't say that the special-casing of enum handling by multiselect is *un*intentional, I do think that the changes were made hastily while failing to consider similar widgets. The multiselect handling was added in order to fix https://github.com/streamlit/streamlit/issues/4909, and because of the reactive nature of the fix, I guess the observation that the behavior really should be generalized to work similarly with all widgets that take a list of options slipped by.

I think what I'll do is spend some time next sprint (which kicks off on Monday) seeing what we can do to generalize

Reply:
I don't mind having to wait a bit, so don't worry about that. But I do think I'd prefer a solution to this issue that keeps intact the invariant we had before this fix; that what is returned by these selector widgets is either a member of the `Iterable`, or `None`. It makes them a whole lot easier to reason about. I hope that's possible to achieve. Saying that, I haven't looked into the motivation for the string casting.

Reply:
In fact, I just noticed that this casting to string was a breaking change for us. We had previously used `multiselect` to select `Enum` values, and the change to casting the value to `str` broke downstream logic that was expecting an `Enum`.

Reply:
Argh, do you happen to have a code snippet that previously worked but is now broken by the new behavior? I have some time to look into fixing this today/tomorrow, and I would have guessed from #4909 that using enums + multiselect was previously broken in general.

Reply:
Ah, nevermind -- I'm guessing the difference is that the example in #4909 defines the enum in the script itself, so the memory addresses of the variants of the enum differ from script run to script run. In a codebase where enums are defined in a separate module, this doesn't end up being an issue since the memory addresses remain stable.

Reply:
> In a codebase where enums are defined in a separate module, this doesn't end up being an issue since the memory addresses remain stable.

Our `Enum` is defined in a separate module, so what you're saying seems to make sense. Let me know if you need more detail.

I guess the question boils down to what to do with `Enums` originating in the script itself. Might have a think about it if you don't arrive at solution. Got a lot of work for the next couple of weeks, so it might be some time before I can look in to it, though.

Reply:
Hi Harahu, thanks for this PR. I have created a PR that will get reviewed ASAP and then after that is merged, we can go ahead with this PR. My current attempt is to inform the user that hey u should create a separate file because doing comparisons / checking for an enum within the returned list will actually fail and cause confusion for the user. Let's see what Vincent thinks of the PR tomorrow. 

Thanks again!

PR for convenience sake: https://github.com/streamlit/streamlit/pull/5224

Reply:
Great effort on this, @willhuang1997! I added my thoughts on a review of the PR. I think it looks really promising.